MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+        var slicedLine: Turf.LineString?
+        
+        /**
+        Implement the Turf.lineSliceAlong(lineString, startDistance, stopDistance) to return a sliced lineString.
+        */
+        if let lineString = currentStepProgress.step.shape {
+            if let midPoint = lineString.coordinateFromStart(distance: startDistance) {
+                slicedLine = lineString.trimmed(from: midPoint, distance: stopDistance - startDistance)
+            }
+        }
+        
+        if slicedLine != nil {
+            return slicedLine!.coordinates.count - 1
+        } else {
+            return 0
+        }
+    }
+
+    private func calculateRouteGranularDistances(coordinates: [CLLocationCoordinate2D]) -> RouteLineGranularDistances? {
+        if coordinates.isEmpty {
+            return nil
+        } else {
+            return calculateGranularDistances(points: coordinates)
+        }
+    }
+    
+    private func calculateGranularDistances(points: [CLLocationCoordinate2D]) -> RouteLineGranularDistances {
+        var distance = 0.0
+        var indexArray = [RouteLineDistancesIndex?](repeating: nil, count: points.count)
+        for index in stride(from: points.count - 1, to: 0, by: -1) {
+            let curr = points[index]
+            let prev = points[index - 1]
+            distance += calculateDistance(point1: curr, point2: prev)
+            indexArray[index - 1] = RouteLineDistancesIndex(point: prev, distanceRemaining: distance)
+        }
+        indexArray[points.count - 1] = RouteLineDistancesIndex(point: points[points.count - 1], distanceRemaining: 0.0)
+        return RouteLineGranularDistances(distance: distance, distanceArray: indexArray.compactMap{ $0 })
+    }
+
+    /**
+    Calculates the distance between 2 points using [EPSG:3857 projection](https://epsg.io/3857).
+    */
+    private func calculateDistance(point1: CLLocationCoordinate2D, point2: CLLocationCoordinate2D) -> Double {
+        let distanceArray: [Double] = [
+            (projectX(point1.longitude) - projectX(point2.longitude)),
+            (projectY(point1.latitude) - projectY(point2.latitude))
+        ]
+        return (distanceArray[0] * distanceArray[0] + distanceArray[1] * distanceArray[1]).squareRoot()
+    }
+
+    private func projectX(_ x: Double) -> Double {
+        return x / 360.0 + 0.5
+    }
+    
+    private func projectY(_ y: Double) -> Double {
+        let sinValue = sin(y * Double.pi / 180)
+        let newYValue = 0.5 - 0.25 * log((1 + sinValue) / (1 - sinValue)) / Double.pi
+        if newYValue < 0 {
+            return 0.0
+        } else if newYValue > 1 {
+            return 1.1
+        } else {
+            return newYValue
+        }
+    }
+    
+    /**
+     Updates the route line appearance from the origin point to the indicated point
+     - parameter point: current position of the puck
+     */
+    public func updateTraveledRouteLine(point: CLLocationCoordinate2D) {
```

Comment:
I don't think that this function should be part of public API, so `public` should be removed.
```suggestion
    func updateTraveledRouteLine(_ coordinate: CLLocationCoordinate2D) {
```

Reply:
Thanks for the suggestion, I'll change the function scope and the parameter name.